### PR TITLE
[mlir][emitc] Lower arith.andi, arith.ori, arith.xori to EmitC

### PR DESCRIPTION
**Pull request by cferry-AMD**
These arith ops have a pretty straightforward lowering to EmitC. This PR also brings in `adaptValueType` and `adaptIntegralTypeSignedness` that other ArithToEmitC functions can benefit from.

### CODE REVIEW

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 1]
```diff
@@ -40,6 +40,29 @@ class ArithConstantOpConversionPattern
...
+                            : IntegerType::SignednessSemantics::Signed;
+      return IntegerType::get(ty.getContext(), ty.getIntOrFloatBitWidth(),
+                              signedness);
+    }
+  }
+  return ty;
+}
+
+/// Insert a cast operation to type \p ty if \p val
+/// does not have this type.
```
Comment (simon-camp) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:61]: Can you reformat this. Looks like it fits into one line.

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 2]
```diff
@@ -40,6 +40,29 @@ class ArithConstantOpConversionPattern
   }
 };
 
+/// Check if the signedness of type \p ty matches the expected
+/// signedness, and issue a type with the correct signedness if
+/// necessary.
```
Comment (simon-camp) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:45]: Reformat

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 3]
```diff
@@ -40,6 +40,29 @@ class ArithConstantOpConversionPattern
...
+Type adaptIntegralTypeSignedness(Type ty, bool needsUnsigned) {
+  if (isa<IntegerType>(ty)) {
+    // Turns signless integers into signed integers.
```
Comment (simon-camp) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:48]: I don't understand this comment. The code seems to suggest something like "Treats singless integers as signed integers"?

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 4]
```diff
@@ -88,6 +88,44 @@ func.func @arith_index(%arg0: index, %arg1: index) {
...
+
+  return
+}
+
+// -----
+
+// CHECK-LABEL: arith_bitwise_bool
+func.func @arith_bitwise_bool(%arg0: i1, %arg1: i1) {
+  // CHECK: %[[AND:[^ ]*]] = emitc.bitwise_and %arg0, %arg1 : (i1, i1) -> i1
+  %5 = arith.andi %arg0, %arg1 : i1
```
Comment (simon-camp) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:118]: Renumber the results from 0 to 2

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 5]
```diff
@@ -88,6 +88,44 @@ func.func @arith_index(%arg0: index, %arg1: index) {
...
 // -----
 
+// CHECK-LABEL: arith_bitwise
+// CHECK-SAME: %[[ARG0:.*]]: i32, %[[ARG1:.*]]: i32
+func.func @arith_bitwise(%arg0: i32, %arg1: i32) {
+  // CHECK: %[[C1:[^ ]*]] = emitc.cast %[[ARG0]] : i32 to ui32
+  // CHECK: %[[C2:[^ ]*]] = emitc.cast %[[ARG1]] : i32 to ui32
+  // CHECK: %[[AND:[^ ]*]] = emitc.bitwise_and %[[C1]], %[[C2]] : (ui32, ui32) -> ui32
+  // CHECK: %[[C3:[^ ]*]] = emitc.cast %[[AND]] : ui32 to i32
+  %5 = arith.andi %arg0, %arg1 : i32
```
Comment (simon-camp) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:98]: Renumber the results from 0 to 2

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 6]
```diff
@@ -88,6 +88,44 @@ func.func @arith_index(%arg0: index, %arg1: index) {
...
+  // CHECK: %[[C3:[^ ]*]] = emitc.cast %[[XOR]] : ui32 to i32
+  %7 = arith.xori %arg0, %arg1 : i32
+
+  return
+}
+
+// -----
+
+// CHECK-LABEL: arith_bitwise_bool
+func.func @arith_bitwise_bool(%arg0: i1, %arg1: i1) {
```
Comment (simon-camp) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:116]: Can you capture the arguments here too. We are not consistent in this file, but we should at least be consistent in one commit.

**Review by simon-camp — CHANGES_REQUESTED** [review]

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 3 above)
Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:48]: This comment meant that signless integers weren't preserved. We however don't need to say anything about signless integers if we know the spec of the function is to get either a signed or an unsigned type.

Comment (mgehre-amd) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp]: In the future, we should forbid using signless integers in the emitc dialect, but right now we still need to handle them.

Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp]: They're still allowed, and the new comment is clearer on what happens to them: we get either a signed, or an unsigned equivalent type; shall we mention that the input may be signed, unsigned or signless?

Comment (mgehre-amd) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp]: For me it's fine, thanks!

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 7]
```diff
@@ -265,6 +288,46 @@ class IntegerOpConversion final : public OpConversionPattern<ArithOp> {
...
+    }
+
+    // Bitwise ops can be performed directly on booleans
+    if (type.isInteger(1)) {
+      rewriter.replaceOpWithNewOp<EmitCOp>(op, type, adaptor.getLhs(),
+                                           adaptor.getRhs());
+      return success();
+    }
+
+    // Bitwise ops are defined by the C standard on unsigned operands.
```
Comment (cferry-AMD) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:310]: For this, see https://learn.microsoft.com/en-us/cpp/c-language/c-bitwise-operators?view=msvc-170.

**Review by simon-camp — APPROVED** [review]
Thanks

**Pull request merged**